쥬스메이커 [STEP2] vetto, 레옹아범

fatherLeon opened the pull request:
@llingo
안녕하세요 링고 거의 일주일만에 다시 PR을 보내네요😅
이번 STEP은 처음 앱을 만드는 과정이고 여러가지 시행착오를 많이 겪다보니 궁금한 내용도 많고 여러가지 생각한 내용과 방법들이 많아서 너무 두서없이 작성한 PR내용도 좀 많은 것 같습니다.
이번주도 지난주처럼 좋은 리뷰 부탁드립니다.🙏🙏

## STEP2 내용 및 핵심경험

* ‘재고수정’ 버튼을 터치하면 ‘재고 추가’ 화면으로 이동합니다.
    1. 이동 방식은 어떤 방식이 적절할까요?
* 각 주문 버튼 터치 시
    1. 쥬스 재료의 재고가 있는 경우 : 쥬스 제조 후 “*** 쥬스 나왔습니다! 맛있게 드세요!” 얼럿 표시
    2. 쥬스 재료의 재고가 없는 경우 : “재료가 모자라요. 재고를 수정할까요?” 얼럿 표시
        1. ‘예’ 선택시 재고수정화면으로 이동
        2. ‘아니오’ 선택시 얼럿 닫기
* 과일쥬스를 제조하여 과일의 재고가 변경되면 화면의 적절한 요소에 변경사항을 반영합니다.
* 이번 스텝 수행 중 핵심 경험
     1. 내비게이션 바 및 바 버튼 아이템의 활용
     2. 얼럿 컨트롤러 활용
     3. Modality의 활용
     4. 화면 사이의 데이터 공유

## 해결하지 못한 점

### 데이터 업데이트
* 저희 코드에서 Label들을 업데이트 해주는 함수가 있습니다. 업데이트를 하게 될 때 변경되지 않은 데이터들까지 불필요하게 변경해주는 것 같습니다.
```swift
func updateLabel(juice: [Fruit: Int]) {
    guard let strawberry = juice[.strawberry],
          let banana = juice[.banana],
          let kiwi = juice[.kiwi],
          let pineapple = juice[.pineapple],
          let mango = juice[.mango] else { return }
        
    strawberryLabel.text = String(strawberry)
    bananaLabel.text = String(banana)
    kiwiLabel.text = String(kiwi)
    pineappleLabel.text = String(pineapple)
    mangoLabel.text = String(mango)
}
```
* 이러한 점은 step2에서는 FruitStockViewController화면에서 재고변경이 이루어지지 않아 아직까진 문제가 없다고 판단했습니다. 하지만 step3에서 재고변경이 이루어 진다면 기능 저하를 만들어낼 문제라고 생각하여 해결하지 못한 점으로 남겼습니다.
* 이런 기능저하를 방지하고자 하나의 label만 변경하는 함수를 만들어서 사용하는게 맞을지? 실제로 대부분의 앱에서 여러 데이터중 하나의 데이터만 변경이 되면 하나의 데이터만 변경을 해주는지? 아니면 전부 다 변경을 해주는지에 대해 의문이 있습니다.

## 고민한점 && 조언을 얻고 싶은 부분
### 화면 사이의 데이터 공유
* 화면 사이의 데이터를 공유하는 방법(다음 뷰컨트롤러의 인스턴스에 할당, 싱글톤, KVO, NotificationCenter 등)에는 여러가지가 있지만 저희가 생각한 방법은 NotificationCenter 방식이랑 singleton 패턴을 사용하는 방식 두가지였습니다. 다만 아직 Notification방식과 다음 뷰컨트롤러에 인스턴스를 할당하는 방식에는 익숙치 않아서 singleton패턴을 사용하여 데이터 공유를 하였습니다.
* 싱글톤을 씀으로 기존 코드에 많은 것을 추가하지 않아도 쉽게 사용가능하고 코드도 짧기 때문에 우선적으로 사용했습니다만, FruitStore의 추가 인스턴스를 못만든다는 점, 각각 클래스가 하나의 인스턴스를 바라보기 때문에 의존성이 높아진다는 단점이 있는데 혹시 어떤 방법을 가장 많이 사용할까요?
* 우선 위 여러 방법(다음 뷰 컨트롤러에 인스턴스 할당, 싱글톤, KVO, NotificationCenter)중에 혹시 어떤 방법을 추천해주실 수 있을까요?


### 화면 전환 방식
* 화면 전환 방식 중 Navigation과 Modal방식에 대해 고민했습니다. 우선 Modal은 사용자에게 순간 집중을 할 수 있게 해주는 요소이고, Navigation은 정보의 깊이를 표현할 때 좋은 화면전환 방식이라고 생각해 Navigation방식을 사용했습니다.
* 이런 화면전환에 대한 고민이 끝나고 난 이후에는 어떻게 구현할 것인지에 대해 많은 고민을 했습니다.
* 스토리보드에서 Segue를 구현하는 방법, 코드로 Segue를 구현하는 방법, alertController를 보여주는 것 처럼 단순하게 present하는 방법, navigationController를 사용하여 pushViewController를 사용하는 방법등에 관해 고민했습니다. 다만, 이미 UI가 구성되어 있고 NavigationController가 스토리보드 내에 있었기 때문에 이 중 두가지 방법을 사용했습니다.
* 우선 메인화면의 ```재고수정```버튼의 경우 간단하고 스토리보드로 한눈에 볼 수 있게 스토리보드 내에서 Segue로 구현하였습니다. 
* 그 외에 실패alert가 나타나서 예를 누를 경우
```swift
let okAction = UIAlertAction(title: "네", style: .default) { (action) in
    guard let fruitStockViewController = self.storyboard?.instantiateViewController(withIdentifier: "FruitStockViewController") else { return }
    self.navigationController?.pushViewController(fruitStockViewController, animated: false)
}                                                        
```
* 다음과 같이 FruitStockViewController의 인스턴스를 받아와 Navigation컨트롤러를 이용하는 pushViewController를 사용하는 방법을 택했습니다.
* 이런경우 두가지 방법으로 하나의 결과(FruitStockViewController를 출력)를 내는 방법으로 작성되어 있는데, 이것을 통일시켜 줘야할까요? 아니면 NavigationController를 사용하여 Navigation방식으로 화면전환한다는 결과는 같기 때문에 굳이 통일시켜줄 필요가 없을까요?
* 스토리보드 내에서 Segue를 사용한 큰 이유는 앱이 어떤흐름을 가지는지 볼 수 있다는 큰 장점이 있어서 사용했고 alertAction내에서는 스토리보드로 만들 수 없기 때문에 위와 같이 사용을 했는데 어떻게 생각하시나요?


### 뷰의 생명주기
```swift
override func viewWillAppear(_ animated: Bool) {
    super.viewWillAppear(animated)
        
    updateLabel(juice: juiceMaker.getFruitsInStore())
}
```
* 재고수정(FruitStockViewController)에서 메인화면(ViewController)로 넘어갈 경우 데이터는 업데이트 되었지만 UILabel에는 업데이트 되지않아서 관련된 방법을 찾아보다가 뷰의 생명주기에 대해 보게 되었습니다.
* viewDidLoad()메소드는 뷰가 처음 Load되고 나서만 실행이 되고, 만약 재고수정(FruitStockViewController)에서 메인화면으로 다시 이동한다면 이미 Load된 상태여서 동작이 안한다는 걸 봤고, [viewWillAppear](https://developer.apple.com/documentation/uikit/uiviewcontroller/1621510-viewwillappear)과 관련해서 뷰가 계층에 추가되는 시점?, 뷰가 보여지기 직전에 호출된다는 것을 보고 사용하였습니다.


### 최대한 적은 수의 함수 사용하기
* 현재 메서드 구현을 살펴보면 하나로 묶여서 실행하는 것을 볼 수 있습니다.
* 위 함수를 작성 전 처음에는 각 버튼마다 "@IBAction" 메소드를 만들어서 실행시키려고 했습니다.
* 그런데 그럴 경우 같은 동작을 하는 함수임에도 불구하고 여러개가 있는게 가독성이나 코드 길이나 너무 불필요하다고 생각하여 이 부분을 해결할 방법을 찾았습니다.
* UIButton은 라벨이 있다면 titleLabel?.text로 해당 텍스트를 불러 올 수 있다는 것을 보고 쥬스를 만드는 모든 버튼을 연결하여 사용하기로 했습니다.
* 이러한 경우 기능성, 가독성 측면에서는 좋아보일 수 있으나 나중에 과일이 추가 되거나 버튼의 기능이 변경되면 리팩토링하는데 어려움을 겪을것 같습니다.
* 단순히 버튼의 라벨 텍스트를 불러와서 사용을 했지만 만약 버튼에 라벨텍스트가 없다면 viewController처럼 고유 ID를 사용하여 쓰는게 좋은지 아니면 그 외 다른 방법이 있는지 궁금합니다

```swift
@IBAction func clickButton(_ sender: UIButton) {
    guard let buttonLabel = sender.titleLabel?.text,
    let juice = decideOrderJuice(buttonLabel) else {
        return
    }
        
    tryMakeJuice(juice)
}
func decideOrderJuice(_ juice: String) -> Juice? {
    switch juice {
    case "딸바쥬스 주문":
        return .strawberryBananaJuice
    case "딸기쥬스 주문":
        return .strawberryJuice
    case "바나나쥬스 주문":
        return .bananaJuice
    case "파인애플쥬스 주문":
        return .pineappleJuice
    case "망키쥬스 주문":
        return .mangoKiwiJuice
    case "키위쥬스 주문":
        return .kiwiJuice
    case "망고쥬스 주문":
        return .mangoJuice
    default:
        return nil
    }
}
```
* 만약 분리를 한다면 이렇게 아래의 코드같이 구현할 것 같습니다.
```swift
@IBAction func clickStrawberryOrderButton() {
    // 기능 구현
}
@IBAction func clickStrawberryBananaOrderButton() {
    // 기능 구현
}
@IBAction func clickPineappleOrderButton() {
    // 기능 구현
}
@IBAction func clickKiwiOrderButton() {
    // 기능 구현
}
```

fatherLeon committed:
Fix: Error, JuiceMaker오류처리 수정

Commit:
feat: 과일 개수 초기화 변경

FruitStore init함수 인자 변, convenience init 함수 삭제

feat

fatherLeon committed:
docs: README 생성

Commit:
feat: 네비게이션 이동 구현

fatherLeon committed:
feat: decrease, increase Fruit 기본 매개변수 값 추가

fatherLeon committed:
docs: README 작성

fatherLeon committed:
docs: README 수정

fatherLeon committed:
docs: README 목차 수정

fatherLeon committed:
docs: README 트러블슈팅 추가

fatherLeon committed:
feat: Juice, JuiceMaker, FruitStore 기능추가

Juice열거형에 name프로퍼티 추가
JuiceMaker에 getFruitsInStore 메소드 추가
FruitStore getFruits 메소드 추가

feat

Commit:
feat: updateLabel 함수 구현

updateLabel 함수 구현

feat

fatherLeon committed:
feat: Alert메소드 구현

쥬스 제조 성공시 나오는 showSuccessAlert메소드
쥬스 제조 실패시 나오는 showFailureAlert메소드 생성 및 구현

feat

Commit:
fear: clickButton 함수 구현

각 음료 버튼 클릭시 해당 쥬스 만드는 함수 구현

feat

fatherLeon committed:
feat: 메소드 수정

JuiceMaker의 makeJuice 반환값 생성
ViewController의 clickButton 수정

feat

Commit:
feat: alert 화면 이동 구현

showFailureAlert함수 내에서 아니오 누를시 재고 수정 화면으로 전환 구현

feat

Commit:
chore: test

fatherLeon committed:
chore: 변수명 수정

fruitStockVC -> fruitStockViewController로 수정

chore

Commit:
feat: 메서드 구현

decideOrderJuice메서드 , tryMakeJuice메서드 구현

feat

fatherLeon committed:
feat: 싱글톤 패턴 구현

FruitStore싱글톤 패턴으로 구현

feat

Commit:
chore: 파일 헤더 주석 변경

AppDelegate, SceneDelegate, ViewController, FruitStockViewController 파일 헤더 주석 변경

chore

fatherLeon committed:
chore: 테스트용 코드 삭제

seu0313 commented on JuiceMaker/JuiceMaker/Model/FruitStoreError.swift:
```diff
@@ -3,7 +3,7 @@
 import Foundation
 
-enum FruitStoreError: Error, LocalizedError {
+enum FruitStoreError: LocalizedError {
     case lackedInventory(fruit: Fruit, lackedAmount: Int)
     case noExistInventory
```
좋습니다 👍

seu0313 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -0,0 +1,35 @@
+    @IBOutlet weak var bananaLabel: UILabel!
+    @IBOutlet weak var pineappleLabel: UILabel!
+    @IBOutlet weak var kiwiLabel: UILabel!
+    @IBOutlet weak var mangoLabel: UILabel!
```
[P1] FruitStockViewController의 UI 요소들을 외부에서 접근할 수 있을 것 같습니다. 
접근제어자를 추가하면 좋을 것 같네요 :)

gzzjk159 replied:
리뷰어 조언대로 외부에서 접근하지 못하고 클래스 내부에서만 접근하도록 `private`를 붙여줬습니다. ec87db366b7548a1349a89c3020738b637ba22d6

seu0313 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -0,0 +1,35 @@
+    @IBOutlet weak var kiwiLabel: UILabel!
+    @IBOutlet weak var mangoLabel: UILabel!
+    
+    private let fruitStore = FruitStore.shared
```
FruitStore를 싱글톤을 사용해서 구현하신 이유가 있을까요? 
싱글톤을 사용 시의 장단점과 주의해야할 점에 대해 고민해보시는 것이 좋을 것 같습니다.

gzzjk159 replied:
싱글톤 패턴은 대규모 프로젝트에는 맞지 않을 수 있지만 쥬스 메이커 같은 작은 프로젝트에선 싱글톤을 사용해도 괜찮다고 생각하였고 과일의 양을 ViewController와 FruitStockViewController에서 공용으로 사용해야 해서 싱글톤을 사용하였습니다. 
장접으로는 데이터 객체가 단 하나만 생성되어 메모리 관리 및 데이터 공유가 쉽다는 장점이 있습니다.
단점으로는 참조가 많아지면 특정 문제가 발생했을때 문제점을 찾기 힘들고 그로 인해 테스트와 수정이 힘듭니다.

seu0313 replied:
싱글톤의 장단점을 잘 이해하고 계시군요 좋습니다. 👍

seu0313 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -0,0 +1,35 @@
+        updateLabel(juice: fruitStore.getFruits())
+    }
+    
+    func updateLabel(juice: [Fruit: Int]) {
```
[P1] 외부에서 updateLabel 메서드에 접근 및 호출할 수 있을 것 같아요

gzzjk159 replied:
리뷰어 조언대로 외부에서 접근하지 못하고 클래스 내부에서만 접근하도록 private를 붙여줬습니다. ec87db366b7548a1349a89c3020738b637ba22d6

seu0313 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -0,0 +1,35 @@
+              let pineapple = juice[.pineapple],
+              let mango = juice[.mango] else { return }
+        
+        strawberryLabel.text = String(strawberry)
```
[P3] String도 있지만 String Interpolation을 사용할 수도 있습니다.

```swift
strawberryLabel.text = "\(strawberry)"
```

gzzjk159 replied:
string interpolation을 사용하게 되면 명료성이 확보가 되면서 코드의 가독성을 높여준다고 생각했습니다. 다른 장점으로는 메모리 낭비를 방지할 수 있다 정도인데 맞는 것인지 잘 모르겠습니다. 단점으로는 포멧을 정해줄수 없다 정도가 있다고 생각합니다.
혹시 다른 장단점을 찾아볼 수 있는 링크가 있는지 궁금합니다.

seu0313 replied:
만약, 문자열을 추가해야하는 경우 String Concatenating보다 Interpolation이 여러 문자열을 합칠때 속도면에서도 더 빠른 이점이 있습니다.

[Swift Evolution SE-0228](https://github.com/apple/swift-evolution/blob/main/proposals/0228-fix-expressiblebystringinterpolation.md)
[Concatenating strings in Swift: Which way is faster?](https://www.globalnerdy.com/2016/02/03/concatenating-strings-in-swift-which-way-is-faster/)

seu0313 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -1,18 +1,106 @@
+    @IBOutlet weak var bananaLabel: UILabel!
+    @IBOutlet weak var kiwiLabel: UILabel!
+    @IBOutlet weak var pineappleLabel: UILabel!
+    @IBOutlet weak var mangoLabel: UILabel!
```
[P1] ViewController 전반적으로 접근제어자를 수정하면 좋을 것 같아요. 코멘트는 여기에만 달아두었습니다.

gzzjk159 replied:
리뷰어 조언대로 외부에서 접근하지 못하고 클래스 내부에서만 접근하도록 private를 붙여줬습니다. ec87db366b7548a1349a89c3020738b637ba22d6

seu0313 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -1,18 +1,106 @@
+        let alert = UIAlertController(title: "재료가 모자라요 재료를 수정할까요?", message: nil, preferredStyle: .alert)
+        let okAction = UIAlertAction(title: "네", style: .default) { (action) in
+            guard let fruitStockViewController = self.storyboard?
+                .instantiateViewController(withIdentifier: "FruitStockViewController") else { return }
```
[P3] 현재는 identifier를 수동으로 입력하도록 되어있는데 만약 철자가 잘못되면 관련된 곳에서 모두 수정해야하는 문제가 있을 것 같아요. FruitStockViewController의 identifier를 FruitStockViewController에서 작성할 수 있도록하면 좋을 것 같아요.

```swift
guard let fruitStockViewController = self.storyboard?.instantiateViewController(
    withIdentifier: FruitStockViewController.identifier) else { ... }
```

---

더 좋은 방법은 `self.storyboard.instantiateViewController(withIdentifier:)`을 간단하게 처리할 수 있는 Sugar를 만드는 방법도 있을 것 같아요.

gzzjk159 replied:
FruitStockViewController에서 `static`를 붙여 `static let identifier = "FruitStockViewController"`를 선언하고 이 전역변수를 호출하여 이동하는 코드로 바꿨습니다. d92cb3a6dcf180bfe330fd08a8c0471eeaf51ca3

seu0313 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -3,27 +3,29 @@
-        self.fruits[.strawberry] = strawberry
-        self.fruits[.banana] = banana
-        self.fruits[.kiwi] = kiwi
-        self.fruits[.mango] = mango
-        self.fruits[.pineapple] = pineapple
+    private init(amount: Int) {
```
> FruitStockViewController와 동일한 코멘트입니다.

FruitStore를 싱글톤을 사용해서 구현하신 이유가 있을까요? 
싱글톤을 사용 시의 장단점과 주의해야할 점에 대해 고민해보시는 것이 좋을 것 같습니다.

fatherLeon replied:
* 우선은 NotificationCenter와 싱글톤 두개를 학습활동시간에 배웠기 때문에 두가지 중 하나를 고민했습니다.
* 처음 STEP1을 할 당시에는 싱글톤을 구현하지 않았고 NotificationCenter역시 구현하지 않았기 떄문에 두가지를 구현해봤고 그 중에서 코드상으로 더 간결하고 앱 내에서 데이터 공유가 쉬운 싱글톤을 사용하게 됐습니다
#### 싱글톤 장점
* 하나의 인스턴스만을 사용하여 앱 내 모든곳에 데이터 공유가 가능하다는점
* 이니셜라이저를 private으로 접근제어 하기 때문에 인스턴스를 하나만 만들어 메모리를 아낄 수 있다는 점
#### 싱글톤 단점
* 하나의 인스턴스를 해당 데이터를 사용하는 앱 내 모든곳에서 사용하기 때문에 결합도가 높아진다는 점
* 싱글톤 해당 클래스가 수정이 될 경우 싱글톤 인스턴스가 사용된 모든 곳을 다 수정해야된다는 점

* 데이터 전달을 하지않고 전체적인 앱 내에서 데이터 공유를 위해 싱글톤을 사용하였습니다.
* STEP3에서는 싱글톤이 아닌 위에 알려주신 delegate를 사용하여 데이터를 전달하고 사용하는 방식을 사용해보겠습니다

seu0313 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -1,18 +1,106 @@
+        default:
+            return nil
+        }
+    }
```
[P2] 누른 버튼의 쥬스 타입을 결정하는 로직으로 판단되는데 더 간단하게 줄여 볼 수 있을까요?

fatherLeon replied:
* 해당 부분을 간단하게 하는 방법으로 UIButton의 tag를 사용하여 case문 길이를 줄이 거나
```swift
switch juice {
    case "딸바쥬스 주문": return .strawberryBananaJuice
    case "딸기쥬스 주문": return .strawberryJuice
}
```
* 이런 형태로 코드 줄 수는 간단하게 바꿀수는 있을거 같은데 이런걸 말씀하시는 걸까요?
* 이외에는 코드를 줄일 방법이 생각나지가 않습니다.

seu0313 replied:
Juice 열거형 타입에 [CaseIterable](https://developer.apple.com/documentation/swift/caseiterable) 프로토콜을 채택하는 방법이 있을 것 같습니다.

```swift
private func decideOrderJuice(_ juice: String) -> Juice? {
    return Juice.allCases.first { juice == "\($0.name) 주문" }
}
```

seu0313 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -0,0 +1,35 @@
+
+import UIKit
+
+class FruitStockViewController: UIViewController {
```
[P1] FruitStockViewController을 상속해서 사용하는 타입이 없다면 final 키워드를 붙여주는 것이 좋아요. 
성능상의 이점과 상속을 하지 못하도록 강제하는 장점이 있습니다.

> final 키워드에 대해서는 추후 학습하실 내용입니다. 평소 final 키워드를 붙여주는 습관을 들여두시는 것을 추천드립니다.

fatherLeon replied:
* 최근 구조체와 클래스의 메모리 관련해서 구조체는 -> 스택, 클래스는 -> 힙에 쌓인다는 내용의 학습활동을 했고 WWDC16의 내용을 한번 더 찾아봤을 때 짧게 나마 final 키워드를 사용하 heap에 할당되는 클래스의 크기가 조금 더 줄어들어서 성능상으로 좋다는 이야기를 본 기억이 있습니다.
* 추후 final키워드에 대해 학습을 가지고 한번 더 정리하는 시간을 가지겠습니다!!👍

seu0313 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -1,18 +1,106 @@
+        updateLabel(juice: juiceMaker.getFruitsInStore())
+    }
+    
+    @IBAction func clickButton(_ sender: UIButton) {
```
[P1] clickButton 만으로는 버튼이 많아졌을때 어떤 버튼이 탭되는 메서드인지 구분하기 힘들 것 같네요. 
아래와 같이 보통 많이 쓰는 것 같아요.

```swift
func xxxButtonDidTap(_ sender: UIButton) { ... }
func didTapXxxButton(_ sender: UIButton) { ... }
```

fatherLeon replied:
```swift
@IBAction func clickButton(_ sender: UIButton) {
}
```
* 기존 함수명 clickButton에서

```swift
@IBAction private func didTapJuiceOrderButton(_ sender: UIButton) {
}
```
* didTapJuiceOrederButton으로 수정하였습니다.

seu0313 reviewed (changes requested):
안녕하세요 vetto, 레옹아범
쥬스메이커 프로젝트 STEP 2 수고하셨습니다 :)

---

> 안내사항은 참고용으로 남겨두었습니다.

1. **PR을 보내기 전에는 3가지를 확인해주세요.**
    - `캠핑사이트 - Course - Pull Request 보낼 때` 글을 꼭 확인해주세요.
    - `yagom-academy:main`로 보내고 있는 것은 아닌 지 확인해주세요.
    - 이전 STEP이 Merge된 후 다음 STEP의 PR을 보내주세요.
2. **PR 수정 코멘트에는 P1 - P3의 중요도를 남겨두었습니다. 중요하지 않는 경우 확인했다는 이모지만 남기셔도 됩니다.**
    - [P1] 무조건 수정이 필요합니다.
    - [P2] 다음 스탭에서 수정해도 됩니다.
    - [P3] 꼭 반영하지 않고 참고만 해도 됩니다.
3. **PR 처럼 수정 후에도 DM를 남겨주시면 빠르게 확인하겠습니다.**

---

**Q1. 화면 사이의 데이터 공유**

> 우선 위 여러 방법(다음 뷰 컨트롤러에 인스턴스 할당, 싱글톤, KVO, NotificationCenter)중에 혹시 어떤 방법을 추천해주실 수 있을까요?

A1. Delegate 패턴을 사용하는 화면 간 데이터 전달을 공부해보시는 것을 추천드립니다.

```swift
// AViewController.swift

class AViewController: BDelegate {
    func presentB() {
        let b = BViewController()
        b.delegate = self
        
        present(b)
    }
    
    func update() {
        // B 화면에서 doSomething이 호출될 때 실행할 로직
    }
}

// BViewController.swift

protocol BDelegate: AnyObject {
    func update()
}

class BViewController {
    weak var delegate: BDelegate?
    
    func doSomething() {
        delegate?.update()
    }
}

```


**Q2. 화면 전환 방식**

> 이런경우 두가지 방법으로 하나의 결과(FruitStockViewController를 출력)를 내는 방법으로 작성되어 있는데, 이것을 통일시켜 줘야할까요? 아니면 NavigationController를 사용하여 Navigation방식으로 화면전환한다는 결과는 같기 때문에 굳이 통일시켜줄 필요가 없을까요?
스토리보드 내에서 Segue를 사용한 큰 이유는 앱이 어떤흐름을 가지는지 볼 수 있다는 큰 장점이 있어서 사용했고 alertAction내에서는 스토리보드로 만들 수 없기 때문에 위와 같이 사용을 했는데 어떻게 생각하시나요?

A2. 개인적인 의견으로 저의 경우 UI를 모두 코드 베이스로 작성하고 있기 때문에 위 문제가 발생할 일이 없었습니다만 굳이 통일해줄 필요는 없다고 생각합니다. 다만, 화면을 띄웠을 때 dismiss 해야하는 주체가 띄운 객체여야한다는 것만 잘 지켜주신다면 문제가 없다고 생각되네요.

```swift
// AViewController.swift

class AViewController {
    func presentB() {
        let b = BViewController()
        present(b)
    }
}

// BViewController.swift

class BViewController {
    func dissmissButtonDidTap() {
        self.dismiss(animated: true) // (X)
        self.presentingViewController?.dismiss(animated: true) // (O)
    }
}

```



**Q3. 최대한 적은 수의 함수 사용하기**

> 단순히 버튼의 라벨 텍스트를 불러와서 사용을 했지만 만약 버튼에 라벨텍스트가 없다면 viewController처럼 고유 ID를 사용하여 쓰는게 좋은지 아니면 그 외 다른 방법이 있는지 궁금합니다

A3. 방법은 여러가지가 있을 것 같습니다. 질문을 남겨주신 것처럼 메서드를 분리하거나 지금처럼 Label의 text를 비교하는 방법도 좋은 방법입니다. 추가적으로 tag를 사용하거나 Enum의 생성자를 활용하거나 UIButton은 클래스 타입이므로 인스턴스 비교하는 방법도 있을 것 같아요.

구현 방법에는 정답이 없기 때문에 팀에서 결정한 대로 진행하면 좋을 것 같네요. :)

Commit:
refactor: 은닉화 및 final 클래스 변경

class final 키워드 설정, private은닉화

refactor

fatherLeon committed:
refactor: 코드 수정

String Interpolation사용, FruitStockViewController Identifier static변수로 선언
clickButton -> didTapJuiceOrderButton으로 함수명 변경

refactor

Commit:
refactor: ViewController 파일명 및 class 명 변경

seu0313 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -0,0 +1,37 @@
+    
+    static let identifier = "FruitStockViewController"
+
+    @IBOutlet weak private var strawberryLabel: UILabel!
+    @IBOutlet weak private var bananaLabel: UILabel!
+    @IBOutlet weak private var pineappleLabel: UILabel!
+    @IBOutlet weak private var kiwiLabel: UILabel!
+    @IBOutlet weak private var mangoLabel: UILabel!
```
좋습니다 👍

seu0313 reviewed (approved):
수고하셨습니다 vetto, 레옹아범 :)
몇 가지 추가 피드백의 경우 다음 스탭에서 수정하셔도 될 것 같아요.
다음 스탭 PR 전에 코멘트 확인 부탁드립니다. 🎉
